consensus/service/src/validators.rs

```diff
@@ -236,279 +186,160 @@ pub mod well_formed_tests {
-        );
-
-        // Change the public key so that it is no longer unique.
-        tx.prefix.outputs[0].public_key = tx_out.public_key.clone();
-
-        assert_eq!(
-            Err(TransactionValidationError::ContainsExistingOutputPublicKey),
-            is_well_formed(&tx, &ledger)
-        );
-    }
-
-    #[test_with_logger]
-    /// `is_well_formed` should reject a transaction that contains an invalid proof-of-membership.
-    fn is_well_formed_rejects_missing_input(_logger: Logger) {
+    /// `is_well_formed` should reject a transaction that contains a proof-of-membership with
+    /// highest index outside the ledger, i.e. a transaction "from the future".
+    fn is_well_formed_rejects_excessive_highest_index(_logger: Logger) {
         let mut rng = Hc128Rng::from_seed([77u8; 32]);
 
         let sender = AccountKey::random(&mut rng);
-        let recipient = AccountKey::random(&mut rng);
-
-        // Create a TxOut that contains an invalid proof-of-membership for some TxOut.
+        // let recipient = AccountKey::random(&mut rng);
+
+        // // Create a TxOut that contains a proof-of-membership whose highest index is outside the ledger.
+        // //
+        // // An easy way to do this is to populate two ledgers of different lengths with the same randomness.
+        // let tx = {
+        //     let mut rng = Hc128Rng::from_seed([142u8; 32]);
+        //     let mut future_ledger = create_ledger();
+        //     let n_blocks = 10;
+        //     initialize_ledger(&mut future_ledger, n_blocks, &sender, &mut rng);
         //
-        // An easy way to do this is to initialize a second ledger with different contents, and
-        // create a transaction that is valid with respect to that ledger.
-        let tx = {
-            let mut bizarro_ledger = create_ledger();
-            let n_blocks = 3;
-            initialize_ledger(&mut bizarro_ledger, n_blocks, &sender, &mut rng);
-
-            // Choose a TxOut to spend. Only the TxOut in the last block is unspent.
-            let block_contents = bizarro_ledger.get_block_contents(n_blocks - 1).unwrap();
-            let tx_out = block_contents.outputs[0].clone();
-
-            create_transaction(
-                &mut bizarro_ledger,
-                &tx_out,
-                &sender,
-                &recipient.default_subaddress(),
-                n_blocks + 1,
-                &mut rng,
-            )
-        };
-
-        // Create the "real" ledger.
-        let mut ledger = create_ledger();
-        let n_blocks = 3;
-        initialize_ledger(&mut ledger, n_blocks, &sender, &mut rng);
-
-        assert_eq!(
-            Err(TransactionValidationError::InvalidTxOutMembershipProof),
-            is_well_formed(&tx, &ledger)
-        );
-    }
-
-    #[test_with_logger]
-    /// `is_well_formed` should reject a transaction with a tombstone block that has been exceeded.
-    fn is_well_formed_rejects_past_tombstone_block(_logger: Logger) {
```

Comment:
Checking if the transaction has expired depends on the current ledger state, and so is out-of-scope for the well-formed check.